### PR TITLE
drain script shouldn't fail when there is a stale HAPRroxy pid in pidfile

### PR DESCRIPTION
**Pull request by kecirlotfi**
We encountered an issue during deployment phase. Bosh filed to drain on canary VM. 
```
Task 6224192 | 23:51:20 | L executing pre-stop: ha_proxy_z2/16e2782c-adcd-4e88-b5b4-e064c9911c57 (0) (canary)
Task 6224192 | 23:51:20 | L executing drain: ha_proxy_z2/16e2782c-adcd-4e88-b5b4-e064c9911c57 (0) (canary) (00:00:03)
                        L Error: Action Failed get_task: Task b5e5ccc7-02af-4fc6-508c-f31d95f95749 result: 1 of 2 drain scripts failed. Failed Jobs: haproxy. Successful Jobs: syslog_forwarder.
Task 6224192 | 23:51:22 | Error: Action Failed get_task: Task b5e5ccc7-02af-4fc6-508c-f31d95f95749 result: 1 of 2 drain scripts failed. Failed Jobs: haproxy. Successful Jobs: syslog_forwarder.

Task 6224192 Started  Tue Aug 23 23:50:51 UTC 2022
Task 6224192 Finished Tue Aug 23 23:51:22 UTC 2022
Task 6224192 Duration 00:00:31
Task 6224192 error
Updating deployment:
  Expected task '6224192' to succeed but state is 'error'

Exit code 1

[bosh][INFO] Deprovisioning client 'd.haproxy.deploy.9b17b2'...
error: lifecycle 'deploy' for deployment 'haproxy' failed in action 'deploy'. Action output (can be empty):

exit status 1
```
After running investigation on the canary VM, we noticed that the drain script fails because the file `/var/vcap/sys/run/bpm/haproxy/haproxy.pid` still exists and non empty. 

The script checks the existence of the pidfile and reads its content then kill the child process that belongs the PID read from the pidfile. 

As the file exists and not empty, `haproxy_wrapper_pid=$(pgrep -P "$pid" haproxy_wrapper)` fails with exit code `1` (process doesn't exist), which makes fail the script.

With the fix we propose, `haproxy_wrapper_pid=$(pgrep -P "$pid" haproxy_wrapper || true)`, awe handle the case of a crashed haproxy_wrapper parent process when its pidfile still exists.

### CODE REVIEW

**Comment by plowin** [comment]
Hi @kecirlotfi , please try rebasing this to the current master, we had some issues with the CI which should be good now!

File: spec/haproxy/templates/drain_spec.rb [hunk 1]
```diff
@@ -31,6 +31,13 @@
           fi
 
           pid="$(cat ${pidfile})"
+          if ! [ pgrep -F "${pidfile}" ]; then
+            # In case haproxy_wrapper process is stale, pid_exists will be empty,
```
Comment (plowin) [inline spec/haproxy/templates/drain_spec.rb:35]: There is no variable named `pid_exists`, like this, the comment is not very useful to me

**Comment by peanball** [comment]
@kecirlotfi please keep this in Draft state until you're ready for it to be reviewed.

File: acceptance-tests/drain_test.go [hunk 2]
```diff
@@ -72,4 +74,51 @@ var _ = Describe("Drain Test", func() {
...
+		// Verify that is in a failing state
+		Expect(boshInstances(deploymentNameForTestNode())[0].ProcessState).To(Or(Equal("failing"), Equal("unresponsive agent")))
+
+		closeLocalServer, localPort := startDefaultTestServer()
+		defer closeLocalServer()
+
+		closeTunnel := setupTunnelFromHaproxyToTestServer(haproxyInfo, haproxyBackendPort, localPort)
+		defer closeTunnel()
+
+		By("Waiting monit to report HAProxy is now healthy (due to having a healthy backend instance)")
```
Comment (domdom82) [inline acceptance-tests/drain_test.go:97]:
```suggestion
		By("Waiting for monit to report HAProxy is now healthy (due to having a healthy backend instance)")
```

File: acceptance-tests/drain_test.go [hunk 3]
```diff
@@ -72,4 +74,51 @@ var _ = Describe("Drain Test", func() {
...
+		expect200(http.Get(fmt.Sprintf("http://%s:8080/health", haproxyInfo.PublicIP)))
+
+		By("Sending a request to HAProxy works")
+		expectTestServer200(http.Get(fmt.Sprintf("http://%s", haproxyInfo.PublicIP)))
+
+		// Set a fake PID of the parent process haproxy_wrapper
+		_, _, err := runOnRemote(haproxyInfo.SSHUser, haproxyInfo.PublicIP, haproxyInfo.SSHPrivateKey, "echo 32761 | sudo tee -a /var/vcap/sys/run/bpm/haproxy/haproxy.pid")
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Draining HAproxy, drain script should not fail and HAproxy should still healthy")
```
Comment (domdom82) [inline acceptance-tests/drain_test.go:115]:
```suggestion
		By("Draining HAproxy, drain script should not fail and HAproxy should still be healthy")
```

File: jobs/haproxy/templates/drain.erb [hunk 4]
```diff
@@ -21,9 +21,17 @@ if [[ ! -f ${pidfile} ]]; then
 fi
 
 pid="$(cat ${pidfile})"
+pid_exists=$(ps -p ${pid} | tail -n +2)
+if [ "${pid_exists}" = '' ]; then
```
Comment (domdom82) [inline jobs/haproxy/templates/drain.erb:25]:
```suggestion
if ! ps -p ${pid} > /dev/null; then
```

File: jobs/haproxy/templates/drain.erb [hunk 5]
```diff
@@ -21,9 +21,17 @@ if [[ ! -f ${pidfile} ]]; then
...
+  # In case haproxy_wrapper process is stale, pid_exists will be empty,
+  # the drain job should not fail
+  echo "$(date): There was no process for the recorded haproxy_wrapper PID." >> ${logfile}
```
Comment (domdom82) [inline jobs/haproxy/templates/drain.erb:28]:
```suggestion
  echo "$(date): There was no process for the recorded haproxy_wrapper PID (${pid})." >> ${logfile}
```

**Review by domdom82 — CHANGES_REQUESTED** [review]

File: jobs/haproxy/templates/drain.erb (hunk 4 above)
Comment (domdom82) [reply jobs/haproxy/templates/drain.erb:25]: as discussed

File: spec/haproxy/templates/drain_spec.rb [hunk 6]
```diff
@@ -31,8 +31,7 @@
           fi
 
           pid="$(cat ${pidfile})"
-          pid_exists=$(ps -p ${pid} | tail -n +2)
-          if [ "${pid_exists}" = '' ]; then
+          if ![ ps -p ${pid} >/dev/null ]; then
```
Comment (domdom82) [inline spec/haproxy/templates/drain_spec.rb:34]: this doesn't work.
```
if ! ps -p ${pid} >/dev/null; then
```
you need to lose the brackets as this is not a test anymore. you are now executing a program and testing for its exit code.

File: spec/haproxy/templates/drain_spec.rb [hunk 7]
```diff
@@ -77,8 +76,7 @@
             fi
 
             pid="$(cat ${pidfile})"
-            pid_exists=$(ps -p ${pid} | tail -n +2)
-            if [ "${pid_exists}" = '' ]; then
+            if ![ ps -p ${pid} >/dev/null ]; then
```
Comment (domdom82) [inline spec/haproxy/templates/drain_spec.rb:79]:
```
if ! ps -p ${pid} >/dev/null; then
```

Comment (kecirlotfi) [reply spec/haproxy/templates/drain_spec.rb]: Fixed. Thanks 👍

**Review by domdom82 — APPROVED** [review]

**Pull request merged**